### PR TITLE
Author specific page locking

### PR DESCRIPTION
**Pull request by kaedroho**
This is an initial implementation of [RFC 43](https://github.com/wagtail/rfcs/pull/43).

This changes the behaviour of page locking so that the person who locked the page can still edit it. It also makes separates the "Unlock" permission from "Lock" so that editors can be given permission to lock pages but not unlock pages that are not locked by them.

Note that the RFC is still up for discussion so please put any comments about the overall design of this feature on the [RFC PR](https://github.com/wagtail/rfcs/pull/43).

This feature was sponsored by The Motley Fool

### CODE REVIEW

File: docs/advanced_topics/settings.rst [hunk 1]
```diff
@@ -661,6 +661,17 @@ Customise the behaviour of rich text fields. By default, ``RichTextField`` and `
...
 
+Page locking
+------------
+
+.. versionadded:: 2.8
+
+  The page locking logic has been changed so that pages are still editable
+  by the users who locked them.
+
+The ``WAGTAILADMIN_GLOBAL_PAGE_EDIT_LOCK`` can be set to ``True`` to prevent users
```
Comment (gasman) [inline docs/advanced_topics/settings.rst]: I suggest removing "The".

File: wagtail/admin/menu.py [hunk 2]
```diff
@@ -116,3 +115,4 @@ def get_context(self, request):
 
 admin_menu = Menu(register_hook_name='register_admin_menu_item', construct_hook_name='construct_main_menu')
 settings_menu = Menu(register_hook_name='register_settings_menu_item', construct_hook_name='construct_settings_menu')
+reports_menu = Menu(register_hook_name='register_reports_menu_item', construct_hook_name='construct_reports_menu')
```
Comment (gasman) [inline wagtail/admin/menu.py]: We should document these hooks in docs/reference/hooks.rst.

File: wagtail/admin/views/reports.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    paginate_by = 10
+
+    def get_context_data(self, *args, object_list=None, **kwargs):
+        context = super().get_context_data(*args, object_list=object_list, **kwargs)
+        context['title'] = self.title
+        context['header_icon'] = self.header_icon
+        return context
+
+
+class LockedPagesView(ReportView):
```
Comment (gasman) [inline wagtail/admin/views/reports.py]: Probably worth adding a permission check for the 'unlock any page' permission to this view (in addition to the one to be added to the menu item).

File: wagtail/admin/action_menu.py [hunk 4]
```diff
@@ -66,7 +66,7 @@ def is_shown(self, request, context):
             return context['user_page_permissions'].for_page(context['parent_page']).can_publish_subpage()
         else:  # view == 'edit' or 'revisions_revert'
             return (
-                not context['page'].locked
+                not context['user_page_permissions'].for_page(context['page']).page_locked()
```
[the remark this answers is not in the document]
Comment (gasman) [reply wagtail/admin/action_menu.py]: Marking as resolved, as this probably isn't something we're going to (or need to) solve in this PR.

File: wagtail/admin/views/reports.py (hunk 3 above)
Comment (kaedroho) [reply wagtail/admin/views/reports.py]: Users can unlock pages that they have locked themselves without requiring this permission, should we add that to the check too?

**Comment by gasman** [comment]
Merged (along with fixes from https://github.com/kaedroho/wagtail/pull/54) in e927f6269a10e3b319c90ad4351c015014109437 + parents.

**Pull request merged**